src/com/twitter/intellij/pants/service/project/resolver/PantsSourceRootsExtension.java

```diff
@@ -67,6 +77,68 @@ public void resolve(
 
       createContentRoots(moduleDataNode, targetInfo);
     }
+    installSyntheticModules(executor, projectDataNode, modules);
+  }
+
+  /**
+   * Takes a set of paths. Returns a new set without paths that are children of any other path in the set
+   * Ex. for [A, A/B, A/B/C, B, C/D] the result set is [A, B, C/D]
+   */
+  public static Set<String> removeChildren(Set<String> paths) {
+    if(paths.size() > 0 ) {
+      List<String> sorted = paths.stream().sorted().collect(Collectors.toList());
+      String current = sorted.get(0);
+      List<String> buf = new ArrayList<>();
+      buf.add(current);
+      for (String s : sorted.stream().skip(1).collect(Collectors.toList())) {
+        if (!s.startsWith(current)) {
+          current = s;
+          buf.add(s);
+        }
+      }
+      return new HashSet<>(buf);
+    }
+    return Collections.emptySet();
+  }
+
+  private void installSyntheticModules(
+    @NotNull PantsCompileOptionsExecutor executor,
+    @NotNull DataNode<ProjectData> projectDataNode,
+    @NotNull Map<String, DataNode<ModuleData>> modules
+  ) {
+    Path projectPath = Paths.get(executor.getBuildRoot().getAbsolutePath());
+    Set<String> replaced =
+      executor.getOptions().getSelectedTargetSpecs().stream()
+        .map(x -> x.replaceFirst("/?:.*", ""))
+        .collect(Collectors.toSet());
+    Set<String> replacedWithoutChildren = removeChildren(replaced);
+    Set<String> allContentRoots =
+      modules.values().stream()
+        .flatMap(x -> x.getChildren()
+        .stream())
+        .map(x -> x.getData(ProjectKeys.CONTENT_ROOT))
+        .filter(Objects::nonNull)
+        .map(x -> x.getRootPath())
+        .collect(Collectors.toSet());
+    replacedWithoutChildren.removeAll(allContentRoots);
+
+    for (String path: replaced) {
+      String moduleName = projectPath.relativize(Paths.get(path)).toString().replaceAll("/", "_") + "-project-root";
```

Comment:
Fastpass uses the `-project-root` naming convention but it still feels a bit weird. Any suggestions for a clearer or more self-explanatory name? 🤔

Reply:
I'm not sure. It could be something less commonly used in build systems, so  it's easier to search it it code/issue tracker. So for example "common ancestor"